Add step by step known issues documentation and update documentation (#13571)
dotnet/arcade: Documentation/Projects/Build Analysis/KnownIssues.md

# Known Issues

> “Known Issues” are GitHub issues that are created to report and track known build or test errors.

Reviewer: [nit] Why the change in quotes?

There are two types of Known Issues:

> - **Infrastructure**: An infrastructure issue is an issue that is not specific to your repository and that requires investigation by the engineering services (@dotnet/dnceng)

Reviewer: Wonder if it's worth specifying Infrastructure Known Issue and Repository Known Issue?

Or maybe too much usage of 'Known'?

Author: it's never too much

- **Repository Known Issue**: A repository issue is an issue that occurs in a particular repository and that should be investigated by the repository owners.

> There are two ways to report a Known Issue: one is via Build Analysis and the other one is manually.

Reviewer: What is our preferred method for creating the issues? manually or through build analysis? I *think* we want to steer folks towards using the check for this.

If that's the case, would it make sense to swap the order in which we provide the options, to make the preferred option the first one?

### Filling out Known Issues JSON blob

> 1.  [Fill the "Error message / Error pattern" section](#how-to-fill-out-a-known-issue-error-section). If

Reviewer: [nit] if what? I have to know!

```suggestion
1.  [Fill the "Error message / Error pattern" section](#how-to-fill-out-a-known-issue-error-section).
```

Author: hahahahaha mori!

## How the matching process works between a Known Issue and a build/test error

The Known Issues feature identifies build and test errors and matches them with open Known Issues. It uses `String.Contains` to compare the errors with the “ErrorMessage” property of Known Issues, or regex matching if an “ErrorPattern” property is provided.

The matching process and its limitations depend on the type of error:

- **Build error**: For build errors, the feature searches for matches in two sources:
  - Azure DevOps error messages
  - Build log of failed jobs

> - **Test errors**:  For test errors, the feature analyzes the errors only when the build has 1000 or fewer failing tests. It also limits the analysis to 100 helix logs. This limitation is due to the high cost of analyzing tests and helix logs. The feature uses the following information to find matches:

Reviewer: [nit, sorry Nit]
```suggestion
- **Test errors**:  For test errors, the feature analyzes the errors only when the build has 1000 or fewer failing tests. It also limits the analysis to 100 Helix logs. This limitation is due to the high cost of analyzing tests and Helix logs. The feature uses the following information to find matches:
```

## What happens after creating or updating a Known Issue

> Knonw issues scans all builds from the last 24 hours since the issue was opened or updated with the error message provided. It also scans the builds that fail after the issue is created.

Reviewer: ```suggestion
Known Issues scan all builds from the last 24 hours since the issue was opened or updated with the error message provided. It also scans all builds that fail after the issue is created.
```

Known issues analyzes both infrastructure issues (Known Issues in dotnet/arcade) and repository issues (Known Issues in the pull request’s repository).

> Additioally displays the Known Issues matches in the Build Analysis check. The example below shows that the feature found 4 matches with the issue “Tracking issue for CI build timeouts”.

Reviewer: ```suggestion
Additionally displays the Known Issues matches in the Build Analysis check. The example below shows that the feature found 4 matches for the issue “Tracking issue for CI build timeouts”.
```

![Known match example](./Resources/KnownIssueMatch.png?raw=true)

### Build retry functionality in Known Issues

> Build Analysis can retry a build that fails with an error that matches a Known Issue. To enable this functionality, the Known Issue must have the ‘BuildRetry’ property set to true. This property indicates that the build failure can be solved by retrying the build.

Reviewer: [nit]
```suggestion
Build Analysis can retry a build that fails with an error that matches a Known Issue. To enable this functionality, the Known Issue must have the ‘BuildRetry’ property set to true. This property indicates that the build failure may be solved by retrying the build.
```
